File: apps/android/core/src/test/kotlin/app/trackevolution/core/GearsTest.kt
package app.trackevolution.core

import app.trackevolution.core.model.LapChannels
import app.trackevolution.core.model.SessionChannels
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.decodeFromJsonElement
import kotlinx.serialization.json.jsonObject
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Assertions.assertNull
import org.junit.jupiter.api.Test

/**
 * The port of `test/unit/gears.test.js`, case for case, plus the cross-language
 * pin against `contracts/logic/gears.json`.
 */
class GearsTest {

    /** A lap in 2nd, up to 3rd, a clutch-in blip, 4th, then back down to 3rd. */
    private val gearA = listOf(2.0, 2.0, 2.0, 3.0, 3.0, 3.0, 3.0, 0.0, 4.0, 4.0, 4.0, 4.0, 3.0, 3.0)

    /** rpm climbs before each upshift and drops after it. */
    private val rpmA = listOf(
        5000.0, 6000.0, 7000.0, 4500.0, 5500.0, 6500.0, 7100.0,
        7100.0, 4800.0, 5200.0, 5600.0, 6000.0, 6900.0, 5000.0,
    )

    private fun lap(
        n: Int,
        timeMs: Int = 90_000,
        gear: List<Double>? = null,
        rpm: List<Double>? = null,
        speed: List<Double>? = null,
    ) = LapChannels(n = n, timeMs = timeMs, speed = speed, rpm = rpm, gear = gear)

    // ---- ordinal / fmtRpm --------------------------------------------------

    @Test
    fun `spells gears as ordinals and 0 as no gear`() {
        assertEquals(
            listOf("no gear", "1st", "2nd", "3rd", "4th", "8th"),
            listOf(0, 1, 2, 3, 4, 8).map { Gears.ordinal(it.toDouble()) },
        )
    }

    @Test
    fun `groups thousands without touching the locale`() {
        assertEquals("6,400", Gears.fmtRpm(6400.0))
        assertEquals("1,000", Gears.fmtRpm(999.6))
        assertEquals("12", Gears.fmtRpm(12.0))
    }

    // ---- gearSegments ------------------------------------------------------

    @Test
    fun `cuts a lap into runs of one gear, keeping gear-0 runs`() {
        assertEquals(
            listOf(
                Gears.GearRun(2.0, 0, 2),
                Gears.GearRun(3.0, 3, 6),
                Gears.GearRun(0.0, 7, 7),
                Gears.GearRun(4.0, 8, 11),
                Gears.GearRun(3.0, 12, 13),
            ),
            Gears.gearSegments(gearA),
        )
    }

    @Test
    fun `segments are empty for a missing or empty series`() {
        assertEquals(emptyList<Gears.GearRun>(), Gears.gearSegments(null))
        assertEquals(emptyList<Gears.GearRun>(), Gears.gearSegments(emptyList()))
    }

    // ---- lapShifts ---------------------------------------------------------

    @Test
    fun `reports each step with the rpm at the last sample in the old gear`() {
        assertEquals(
            listOf(
                Gears.Shift(k = 3, from = 2.0, to = 3.0, up = true, rpm = 7000.0),
                // Read at k=6, the last sample in 3rd, not the clutch-in blip.
                Gears.Shift(k = 8, from = 3.0, to = 4.0, up = true, rpm = 7100.0),
                Gears.Shift(k = 12, from = 4.0, to = 3.0, up = false, rpm = 6000.0),
            ),
            Gears.lapShifts(lap(1, gear = gearA, rpm = rpmA)),
        )
    }

    @Test
    fun `skips a clutch-in stretch rather than counting it as a gear`() {
        assertEquals(emptyList<Gears.Shift>(), Gears.lapShifts(lap(1, gear = listOf(3.0, 0.0, 0.0, 3.0))))
    }

    @Test
    fun `gives null rpm without an rpm series and nothing without a gear series`() {
        assertEquals(
            listOf(Gears.Shift(k = 1, from = 2.0, to = 3.0, up = true, rpm = null)),
            Gears.lapShifts(lap(1, gear = listOf(2.0, 3.0))),
        )
        assertEquals(emptyList<Gears.Shift>(), Gears.lapShifts(lap(1, rpm = rpmA)))
    }

    // ---- shiftPoints -------------------------------------------------------

    private val channels = SessionChannels(
        v = 1,
        dStepM = 20.0,
        laps = listOf(
            lap(1, 90_000, gear = gearA, rpm = rpmA),
            lap(2, 89_000, gear = gearA, rpm = rpmA.mapIndexed { k, v -> if (k == 2) 6400.0 else if (k == 6) 7300.0 else v }),
            lap(3, 91_000, gear = gearA), // no rpm: not counted
            lap(4, 92_000, speed = listOf(100.0, 100.0)), // no gear: not counted
        ),
    )

    @Test
    fun `reduces upshifts to min - median - max rpm per gear, downshifts excluded`() {
        val sp = Gears.shiftPoints(channels)!!
        assertEquals(
            listOf(
                Gears.GearShifts(gear = 2.0, count = 2, minRpm = 6400.0, medianRpm = 6700, maxRpm = 7000.0),
                Gears.GearShifts(gear = 3.0, count = 2, minRpm = 7100.0, medianRpm = 7200, maxRpm = 7300.0),
            ),
            sp.gears,
        )
        assertEquals(7050, sp.medianRpm)
        assertEquals(7300.0, sp.maxRpm)
    }

    @Test
    fun `shift points are null when no lap carries gear and rpm, or nothing upshifts`() {
        assertNull(
            Gears.shiftPoints(
                SessionChannels(v = 1, dStepM = 20.0, laps = listOf(channels.laps[2], channels.laps[3])),
            ),
        )
        assertNull(
            Gears.shiftPoints(
                SessionChannels(
                    v = 1,
                    dStepM = 20.0,
                    laps = listOf(lap(1, gear = listOf(3.0, 3.0, 3.0), rpm = listOf(5000.0, 5000.0, 5000.0))),
                ),
            ),
        )
        assertNull(Gears.shiftPoints(SessionChannels(v = 1, dStepM = 20.0, laps = emptyList())))
    }

    // ---- shiftNotes --------------------------------------------------------

    @Test
    fun `names the gears taken to the top of the rev range and the ones shifted early`() {
        val notes = Gears.shiftNotes(
            Gears.SessionShifts(
                gears = listOf(
                    Gears.GearShifts(2.0, 3, 6900.0, 7000, 7150.0),
                    Gears.GearShifts(3.0, 3, 6800.0, 6950, 7100.0),
                    Gears.GearShifts(4.0, 2, 6000.0, 6200, 6400.0),
                    // One shift: not a pattern.
                    Gears.GearShifts(5.0, 1, 5000.0, 5000, 5000.0),
                ),
                medianRpm = 6900,
                maxRpm = 7200.0,
            ),
        )
        assertEquals(
            listOf(
                "Upshifts from 2nd and 3rd run to the top of the rev range seen today (≈7,200 rpm).",
                "Upshifts from 4th come ≈800 rpm earlier than from 2nd.",
            ),
            notes,
        )
    }

    @Test
    fun `says nothing when the gears agree and none reaches the limit`() {
        assertEquals(
            emptyList<String>(),
            Gears.shiftNotes(
                Gears.SessionShifts(
                    gears = listOf(
                        Gears.GearShifts(2.0, 3, 6400.0, 6500, 6600.0),
                        Gears.GearShifts(3.0, 3, 6300.0, 6400, 6500.0),
                    ),
                    medianRpm = 6450,
                    maxRpm = 7200.0,
                ),
            ),
        )
        assertEquals(emptyList<String>(), Gears.shiftNotes(null))
    }

    // ---- gearDisagreements -------------------------------------------------

    @Test
    fun `outlines runs where highlighted laps sit in different gears`() {
        val a = listOf(2.0, 2.0, 3.0, 3.0, 3.0, 3.0, 3.0, 3.0, 4.0, 4.0)
        // 4th where a is in 3rd for four points.
        val b = listOf(2.0, 2.0, 4.0, 4.0, 4.0, 4.0, 3.0, 3.0, 4.0, 4.0)
        assertEquals(listOf(Gears.Disagreement(2, 5)), Gears.gearDisagreements(listOf(a, b)))
    }

    @Test
    fun `ignores a shift that merely lands a sample later, and gear 0 on either lap`() {
        val a = listOf(2.0, 2.0, 3.0, 3.0, 3.0, 3.0, 3.0)
        // One-point offset: a shift, not a choice.
        val b = listOf(2.0, 2.0, 2.0, 3.0, 3.0, 3.0, 3.0)
        assertEquals(emptyList<Gears.Disagreement>(), Gears.gearDisagreements(listOf(a, b)))
        // Clutch in where a is in 3rd: no disagreement.
        val c = listOf(2.0, 2.0, 0.0, 0.0, 0.0, 0.0, 3.0)
        assertEquals(emptyList<Gears.Disagreement>(), Gears.gearDisagreements(listOf(a, c)))
        assertEquals(3, Gears.MIN_DISAGREE_POINTS)
    }

    @Test
    fun `needs two series and honours the run threshold`() {
        assertEquals(
            emptyList<Gears.Disagreement>(),
            Gears.gearDisagreements(listOf(listOf(2.0, 3.0))),
        )
        assertEquals(
            listOf(Gears.Disagreement(1, 2)),
            Gears.gearDisagreements(listOf(listOf(2.0, 3.0, 3.0), listOf(2.0, 4.0, 4.0)), minRun = 1),
        )
    }

    // ---- cross-language pin ------------------------------------------------

    /**
     * The JS implementation's own output for a shared input has to come back out
     * of this port integer for integer and string for string.
     */
    @Test
    fun `matches the JavaScript implementation on a shared fixture`() {
        val json = Json { ignoreUnknownKeys = false }
        val fixture = Json.parseToJsonElement(
            RepoRoot.path("contracts/logic/gears.json").readText(),
        ).jsonObject
        val input = fixture["input"]!!.jsonObject
        val expected = fixture["expected"]!!.jsonObject
        val channels = json.decodeFromJsonElement<SessionChannels>(input["channels"]!!)

        assertEquals(
            json.decodeFromJsonElement<List<List<Gears.GearRun>>>(expected["segments"]!!),
            channels.laps.map { Gears.gearSegments(it.gear) },
        )
        assertEquals(
            json.decodeFromJsonElement<List<List<Gears.Shift>>>(expected["shifts"]!!),
            channels.laps.map { Gears.lapShifts(it) },
        )

        val sp = Gears.shiftPoints(channels)
        assertNotNull(sp)
        assertEquals(json.decodeFromJsonElement<Gears.SessionShifts>(expected["shiftPoints"]!!), sp)
        assertEquals(
            json.decodeFromJsonElement<List<String>>(expected["notes"]!!),
            Gears.shiftNotes(sp),
        )
        assertEquals(JsonNull, expected["noRpm"])
        assertNull(
            Gears.shiftPoints(
                SessionChannels(v = 1, dStepM = 20.0, laps = listOf(channels.laps[2], channels.laps[3])),
            ),
        )

        val a = channels.laps[0].gear
        val b = channels.laps[1].gear
        val c = channels.laps[2].gear
        assertEquals(
            json.decodeFromJsonElement<List<Gears.Disagreement>>(expected["disagreementsAB"]!!),
            Gears.gearDisagreements(listOf(a, b)),
        )
        assertEquals(
            json.decodeFromJsonElement<List<Gears.Disagreement>>(expected["disagreementsABC"]!!),
            Gears.gearDisagreements(listOf(a, b, c)),
        )
        val offsetA = listOf(2.0, 2.0, 3.0, 3.0, 3.0)
        val offsetB = listOf(2.0, 2.0, 2.0, 3.0, 3.0)
        assertEquals(
            json.decodeFromJsonElement<List<Gears.Disagreement>>(expected["offsetDefault"]!!),
            Gears.gearDisagreements(listOf(offsetA, offsetB)),
        )
        assertEquals(
            json.decodeFromJsonElement<List<Gears.Disagreement>>(expected["offsetMinRun1"]!!),
            Gears.gearDisagreements(listOf(offsetA, offsetB), minRun = 1),
        )
    }
}

File: test/unit/gears.test.js
import { describe, expect, it } from "vitest";
import {
  MIN_DISAGREE_POINTS,
  fmtRpm,
  gearDisagreements,
  gearRibbonSvg,
  gearSegments,
  lapShifts,
  ordinal,
  shiftNotes,
  shiftPoints,
  shiftTableHtml,
} from "../../public/js/gears.js";

// A lap in 2nd, up to 3rd, a clutch-in blip, 4th, then back down to 3rd.
const gearA = [2, 2, 2, 3, 3, 3, 3, 0, 4, 4, 4, 4, 3, 3];
// rpm climbs before each upshift and drops after it.
const rpmA = [5000, 6000, 7000, 4500, 5500, 6500, 7100, 7100, 4800, 5200, 5600, 6000, 6900, 5000];

describe("ordinal / fmtRpm", () => {
  it("spells gears as ordinals and 0 as no gear", () => {
    expect([0, 1, 2, 3, 4, 8].map(ordinal)).toEqual(["no gear", "1st", "2nd", "3rd", "4th", "8th"]);
  });
  it("groups thousands without touching the locale", () => {
    expect(fmtRpm(6400)).toBe("6,400");
    expect(fmtRpm(999.6)).toBe("1,000");
    expect(fmtRpm(12)).toBe("12");
  });
});

describe("gearSegments", () => {
  it("cuts a lap into runs of one gear, keeping gear-0 runs", () => {
    expect(gearSegments(gearA)).toEqual([
      { gear: 2, k0: 0, k1: 2 },
      { gear: 3, k0: 3, k1: 6 },
      { gear: 0, k0: 7, k1: 7 },
      { gear: 4, k0: 8, k1: 11 },
      { gear: 3, k0: 12, k1: 13 },
    ]);
  });
  it("is empty for a missing or empty series", () => {
    expect(gearSegments(undefined)).toEqual([]);
    expect(gearSegments([])).toEqual([]);
  });
});

describe("lapShifts", () => {
  it("reports each step with the rpm at the last sample in the old gear", () => {
    expect(lapShifts({ gear: gearA, rpm: rpmA })).toEqual([
      { k: 3, from: 2, to: 3, up: true, rpm: 7000 },
      { k: 8, from: 3, to: 4, up: true, rpm: 7100 }, // read at k=6, the last sample in 3rd, not the clutch-in blip
      { k: 12, from: 4, to: 3, up: false, rpm: 6000 },
    ]);
  });
  it("skips a clutch-in stretch rather than counting it as a gear", () => {
    expect(lapShifts({ gear: [3, 0, 0, 3] })).toEqual([]);
  });
  it("gives null rpm without an rpm series and nothing without a gear series", () => {
    expect(lapShifts({ gear: [2, 3] })).toEqual([{ k: 1, from: 2, to: 3, up: true, rpm: null }]);
    expect(lapShifts({ rpm: rpmA })).toEqual([]);
    expect(lapShifts(null)).toEqual([]);
  });
});

describe("shiftPoints", () => {
  const channels = {
    v: 1,
    dStepM: 20,
    laps: [
      { n: 1, timeMs: 90_000, gear: gearA, rpm: rpmA },
      { n: 2, timeMs: 89_000, gear: gearA, rpm: rpmA.map((v, k) => (k === 2 ? 6400 : k === 6 ? 7300 : v)) },
      { n: 3, timeMs: 91_000, gear: gearA }, // no rpm: not counted
      { n: 4, timeMs: 92_000, speed: [100, 100] }, // no gear: not counted
    ],
  };

  it("reduces upshifts to min / median / max rpm per gear, downshifts excluded", () => {
    const sp = shiftPoints(channels);
    expect(sp.gears).toEqual([
      { gear: 2, count: 2, minRpm: 6400, medianRpm: 6700, maxRpm: 7000 },
      { gear: 3, count: 2, minRpm: 7100, medianRpm: 7200, maxRpm: 7300 },
    ]);
    expect(sp.medianRpm).toBe(Math.round((7000 + 7100) / 2));
    expect(sp.maxRpm).toBe(7300);
  });

  it("is null when no lap carries gear and rpm, or nothing upshifts", () => {
    expect(shiftPoints({ v: 1, dStepM: 20, laps: [channels.laps[2], channels.laps[3]] })).toBeNull();
    expect(shiftPoints({ v: 1, dStepM: 20, laps: [{ n: 1, gear: [3, 3, 3], rpm: [5000, 5000, 5000] }] })).toBeNull();
    expect(shiftPoints(null)).toBeNull();
  });
});

describe("shiftNotes", () => {
  it("names the gears taken to the top of the rev range and the ones shifted early", () => {
    const notes = shiftNotes({
      gears: [
        { gear: 2, count: 3, minRpm: 6900, medianRpm: 7000, maxRpm: 7150 },
        { gear: 3, count: 3, minRpm: 6800, medianRpm: 6950, maxRpm: 7100 },
        { gear: 4, count: 2, minRpm: 6000, medianRpm: 6200, maxRpm: 6400 },
        { gear: 5, count: 1, minRpm: 5000, medianRpm: 5000, maxRpm: 5000 }, // one shift: not a pattern
      ],
      medianRpm: 6900,
      maxRpm: 7200,
    });
    expect(notes).toEqual([
      "Upshifts from 2nd and 3rd run to the top of the rev range seen today (≈7,200 rpm).",
      "Upshifts from 4th come ≈800 rpm earlier than from 2nd.",
    ]);
  });
  it("says nothing when the gears agree and none reaches the limit", () => {
    expect(
      shiftNotes({
        gears: [
          { gear: 2, count: 3, minRpm: 6400, medianRpm: 6500, maxRpm: 6600 },
          { gear: 3, count: 3, minRpm: 6300, medianRpm: 6400, maxRpm: 6500 },
        ],
        medianRpm: 6450,
        maxRpm: 7200,
      })
    ).toEqual([]);
    expect(shiftNotes(null)).toEqual([]);
  });
});

describe("gearDisagreements", () => {
  it("outlines runs where highlighted laps sit in different gears", () => {
    const a = [2, 2, 3, 3, 3, 3, 3, 3, 4, 4];
    const b = [2, 2, 4, 4, 4, 4, 3, 3, 4, 4]; // 4th where a is in 3rd for four points
    expect(gearDisagreements([a, b])).toEqual([{ k0: 2, k1: 5 }]);
  });
  it("ignores a shift that merely lands a sample later, and gear 0 on either lap", () => {
    const a = [2, 2, 3, 3, 3, 3, 3];
    const b = [2, 2, 2, 3, 3, 3, 3]; // one-point offset: a shift, not a choice
    expect(gearDisagreements([a, b])).toEqual([]);
    const c = [2, 2, 0, 0, 0, 0, 3]; // clutch in where a is in 3rd: no disagreement
    expect(gearDisagreements([a, c])).toEqual([]);
    expect(MIN_DISAGREE_POINTS).toBe(3);
  });
  it("needs two series and honours the run threshold", () => {
    expect(gearDisagreements([[2, 3]])).toEqual([]);
    expect(gearDisagreements([[2, 3, 3], [2, 4, 4]], 1)).toEqual([{ k0: 1, k1: 2 }]);
  });
});

describe("gearRibbonSvg", () => {
  const channels = {
    v: 1,
    dStepM: 20,
    laps: [
      { n: 1, timeMs: 90_000, speed: Array(14).fill(100), gear: gearA, rpm: rpmA },
      { n: 2, timeMs: 89_000, speed: Array(14).fill(100), gear: [2, 2, 2, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3] },
      { n: 3, timeMs: 91_000, speed: Array(14).fill(100) }, // no gear
    ],
  };
  const label = (i) => `Lap ${i + 1}`;

  it("draws one block per gear run, a gap for gear 0, on the channel charts' axis", () => {
    const svg = gearRibbonSvg(channels, new Map([[0, "var(--chart-line)"]]), label);
    expect(svg).toContain('data-channel="gear"');
    expect(svg).toContain('data-padl="56"');
    expect(svg).toContain(`data-x1="${13 * 20}"`);
    expect((svg.match(/<rect /g) || []).length).toBe(4); // 2, 3, 4, 3 — the 0 run is a gap
    expect(svg).toContain("Lap 1");
    expect(svg).not.toContain("dashed");
  });

  it("outlines the runs where two highlighted laps disagree", () => {
    const lit = new Map([
      [0, "var(--chart-line)"],
      [1, "var(--chart-line-b)"],
    ]);
    const svg = gearRibbonSvg(channels, lit, label);
    // 4 blocks + 2 blocks + one dashed outline (lap 1 in 4th for k=8..11 while lap 2 stays in 3rd)
    expect((svg.match(/<rect /g) || []).length).toBe(7);
    expect(svg).toContain('stroke-dasharray="3 2"');
    expect(svg).toContain("laps disagree");
  });

  it("returns '' when no highlighted lap has a gear series", () => {
    expect(gearRibbonSvg(channels, new Map([[2, "var(--chart-line)"]]), label)).toBe("");
    expect(gearRibbonSvg(channels, new Map(), label)).toBe("");
  });
});

describe("shiftTableHtml", () => {
  it("tabulates upshift rpm per gear with the notes beneath, and is '' without data", () => {
    const html = shiftTableHtml({
      v: 1,
      dStepM: 20,
      laps: [
        { n: 1, gear: gearA, rpm: rpmA },
        { n: 2, gear: gearA, rpm: rpmA },
      ],
    });
    expect(html).toContain('class="ch-shifts"');
    expect(html).toContain("From 2nd");
    expect(html).toContain("From 3rd");
    expect(html).toContain("7,100"); // 3rd's upshift rpm
    expect(html).toContain("run to the top of the rev range"); // 3rd hits the session max
    expect(shiftTableHtml({ v: 1, dStepM: 20, laps: [{ n: 1, speed: [1, 2] }] })).toBe("");
  });
});
